ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java

```diff
@@ -938,6 +949,339 @@ private void updateClusterMapInHelix(boolean startValidatingClusterManager) thro
+    } else {
+      info("[{}] We don't have any partition to add to existing resources for cluster {}", dcName.toUpperCase(),
+          clusterName);
+    }
+  }
+
+  /**
+   * Create or update ideal state for given partition layout when the resources and static layout are both FULL_AUTO
+   * compatible. When calling this method, make sure that given partitions' instances are not assigned to any existing
+   * resources, so they are all new partitions in new instances.
+   * This method would first find all the instance group and partitions that belong to each group. It will then try to
+   * merge those groups if the size of the group is less than the max number of instances in one resource. Also this
+   * method would try to merge with last existing resource if the size allows.
+   * @param dcName The name of the datacenter
+   * @param partitionsToInstancesInDc The map of partition layout.
+   * @param resourceIdToInstances The map from resource id to a list of instance names that are assigned to this resource
+   * @param resourceIdToIdealState The map from resource id to IdealState
+   */
+  private void createOrUpdateIdealStateForFullAuto(String dcName, Map<String, Set<String>> partitionsToInstancesInDc,
+      TreeMap<Integer, Set<String>> resourceIdToInstances, Map<Integer, IdealState> resourceIdToIdealState) {
+    // The remaining partition would form a list of instance group. Each group would be added to the new partitions
+    int remainingPartition = partitionsToInstancesInDc.size();
+    HelixAdmin dcAdmin = adminForDc.get(dcName);
+    List<Set<String>> instanceGroups = groupInstancesBasedOnSharedPartition(partitionsToInstancesInDc);
+    Map<String, Set<String>> instanceToPartitions = generateInstanceToPartitionsMap(partitionsToInstancesInDc);
+    info("[{}] Found {} groups of instances for cluster {}", dcName.toUpperCase(), instanceGroups.size(), clusterName);
+    Iterator<Set<String>> instanceGroupIterator = instanceGroups.iterator();
+    int startResourceId = FULL_AUTO_COMPATIBLE_RESOURCE_NAME_START_NUMBER;
+    int maxResource = -1;
+    Set<String> currentInstances = new HashSet<>();
+    // The last resource might still be able to take some group
+    if (!resourceIdToInstances.isEmpty()) {
+      Map.Entry<Integer, Set<String>> lastEntry = resourceIdToInstances.lastEntry();
+      startResourceId = lastEntry.getKey();
+      maxResource = startResourceId;
+      currentInstances.addAll(lastEntry.getValue());
+      info("[{}] Already has resources, will try to use last resource {}, current size: {}, target size {}",
+          dcName.toUpperCase(), startResourceId, currentInstances.size(), maxInstancesInOneResourceForFullAuto);
+    }
+    int curResourceId = startResourceId;
+    while (instanceGroupIterator.hasNext()) {
+      Set<String> group = instanceGroupIterator.next();
+      if (currentInstances.size() + group.size() > maxInstancesInOneResourceForFullAuto) {
```

Comment:
It can't. This logic is here to prevent that from happening. If the currentInstances size is 18, and the maxInstance is 18, then we will create a new resource for it in this if statement .